java/arcs/core/host/ArcHostContextParticle.kt

```diff
@@ -104,7 +104,10 @@ class ArcHostContextParticle : AbstractArcHostParticle() {
 
             // construct a map of particleName to Particle instance
             val instantiatedParticles = particleEntities.map {
-                it.particleName to instantiateParticle(ParticleIdentifier.from(it.location))
+                it.particleName to instantiateParticle(
+                    ParticleIdentifier.from(it.location),
+                    arcHostContext.particles[it.particleName]!!.planParticle
```

Comment:
Are we 100% sure arcHostContext will have this information here?

Reply:
I don't know how the whole system works, but here is my reasoning why it should be here: 
We're inside a `map` function of the Entities that represents the context. In theory, `it` represents the same `Plan.Particle` only in an "inflated" form. This gives me confidence that the information will be here at this time. 

Alternatives: 
* Allow the `spec` argument to be nullable
* Provide a default spec (`it` contains plenty of metadata like the particle name and location). 

WDYT, @cromwellian ?

Reply:
The initial ArcHostContext before readArcHostContext is called contains only an ArcId and EntityManager, sufficient to perform all of the operations for deserialization.  It will not have the Plan, that's what it is going to deserialize.  Remember, the ArcHost is waking up a dormant Arc here, so whatever it has to use has to be something that was serialized. 

You can't serialize these spec objects, unless they've been registered in a static registry somewhere, and referred to by name. 

Since you're only doing this to support dynamic EntityBase/EntitySpec,  there's really only one kind of spec right? All others are expected to be from schema2kotlin?

Ergo, I think I would modify AbstractArcHost itself to support this, and change the way ArcHostContextParticle works. I've been meaning to do this anyway. Don't allow ArcHostContextParticle to instantiate particles, rather, make the ArcHostContext particle field a nullable 'var',  and after invoking readArcHostContext(), the AbstractArcHost can then use the newly read Plan.Partition to invoke instantiateParticle, with or without the Plan.Particle information. 

Does that make sense?

Reply:
What we need and what is not serialised is the type of the connection (e.g. `[Product {name: Text, rating: Number}]`). Looking at schemas in ArcHostContext.arcs it stores the Type.TAG - which is what is needed for the storage stack to choose the right CRDT, but we need an actual field names and types of schemas.

You cannot do what we need if this information (field names, field types) is not available at deserialization. My proposal would be to store some serialization of the type (JSON, Proto, `[Product {name: Text, rating: Number}]` - string with custom serializer and deserializer? I don't know) in the HandleConnection schema.

Thoughts @cromwellian ?

Reply:
@piotrswigon I think the ArcHost can get the schema / field information from the Type.Tag via [this function](https://github.com/PolymerLabs/arcs/blob/2cdd8856d01414610bcdaeb765d3f0173535132d/java/arcs/core/host/ArcHostContextParticle.kt#L201) called [here](https://github.com/PolymerLabs/arcs/blob/2cdd8856d01414610bcdaeb765d3f0173535132d/java/arcs/core/host/ArcHostContextParticle.kt#L181).

If I understand correctly, what Ray is proposing is a way to change the ArcHost to get the .SCHEMA without having to re-instantiate the particles on read (specifically, reading the `ArcHostContext`). 

My understanding so far is that the Schemas exist from the generated Kotlin code (`schema2kotlin`), they just need to be referenced in a special way due to how they were generated (the schema gets registered upon class load time, making this a bit tricky).

Reply:
The function you linked is getting the schema from the code generated superclass. Note that we will not have this code generated superclass in the case of your generic egress particle.

The only source of the type information that the generic egress particle has the the Plan.Partition received from the allocator. Once the ArcHost is destroyed and the resurrect call comes back the type information of the particle is no longer there and nowhere to be found.

Options that I see:
1) Egress particle stores the typing itself in a custom handle.
2) We modify the ArcHostContext to have expanded view of the type, beyond Type.TAG.

Reply:
The fundamental problem is that Schema's are not serializable due to the refinement/query fields being lambdas. The long term fix for that and other things, is to move to a DSL/AST representation, this will be needed anyway for efficient queries. 

A temporary workaround for deserializing the arc is to instantiate the particle once with fake information, get the entitySpecs/schema to rehydrate the Plan, and then let ArcHost reinstantiate the particle with the real information. 

This is ugly, but might be the only option until Schema is serializable. 

Alternatively, we could add an extra method to the particle interface, Particle.initialize(Plan.particle), and stick with a no arg ctor. The initialize() method is the "real" ctor.

Reply:
Another short term workaround could be to rely on SchemaRegistry. Instead of serialising the schema, we just store its hash, which we use to look up into the SchemaRegistry at deserialisation time.

Ray, WDYT?

(I don't BTW know what are the long term plans for SchemaRegistry, but it exists as of now)